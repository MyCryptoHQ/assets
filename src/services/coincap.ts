import { Asset } from '../constants';
import { CachedResponse, getInitialCache, withCache } from './cache';
import { isSimilar } from '../utils';

const COINCAP_API_ENDPOINT = 'https://api.coincap.io/v2/assets?limit=2000';

interface CoinCapCoin {
  id: string;
  symbol: string;
  name: string;
}

export interface CoinCapData {
  coins: CoinCapCoin[];
}

/**
 * Match an asset with a CoinCap ID.
 *
 * @param {Partial<CoinCapData>} cache
 * @param {Asset} asset
 * @return {Promise<CachedResponse<CoinCapData, string | null>>}
 */
export const matchCoinCapId = async (
  cache: Partial<CoinCapData>,
  asset: Asset
): Promise<CachedResponse<CoinCapData, string | null>> => {
  const data = cache;
  if (!data || !data.coins) {
    data.coins = (await getInitialCache<{ data: CoinCapCoin[] }>(COINCAP_API_ENDPOINT)).data;
  }

  const match = data.coins.find(coin => coin.symbol === asset.symbol);
  if (match && isSimilar(match.name, asset.name)) {
    return {
      cache: data,
      data: match.id
    };
  }

  return {
    cache: data,
    data: null
  };
};

export default withCache<CoinCapData, typeof matchCoinCapId, string | null>(matchCoinCapId);
