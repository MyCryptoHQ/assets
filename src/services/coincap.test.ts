import { CoinCapData, matchCoinCapId } from './coincap';
import { getInitialCache } from './cache';

const CACHE_DATA = [
  {
    id: 'ethereum',
    symbol: 'ETH',
    name: 'Ethereum'
  }
];

jest.mock('./cache', () => ({
  getInitialCache: jest.fn(async () => ({
    data: CACHE_DATA
  })),

  withCache: jest.requireActual('./cache').withCache
}));

afterEach(() => {
  jest.clearAllMocks();
});

describe('matchCoinCapId', () => {
  it('initializes the cache', async () => {
    const asset = {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    const { cache } = await matchCoinCapId({}, asset);
    const result = await matchCoinCapId(cache, asset);

    expect(result.cache.coins).toMatchSnapshot();
    expect(getInitialCache).toHaveBeenCalledTimes(1);
  });

  it('fetches the ID of an asset', async () => {
    const asset = {
      name: 'Ethereum',
      symbol: 'ETH',
      uuid: '356a192b-7913-504c-9457-4d18c28d46e6'
    };

    await expect(matchCoinCapId({}, asset)).resolves.toEqual({
      cache: {
        coins: CACHE_DATA
      },
      data: 'ethereum'
    });
  });

  it('returns null if an asset could not be found', async () => {
    const asset = {
      name: 'Ethereum Classic',
      symbol: 'ETC',
      uuid: '6c1e671f-9af5-546d-9c1a-52067bdf0e53'
    };

    await expect(matchCoinCapId({}, asset)).resolves.toEqual({
      cache: {
        coins: CACHE_DATA
      },
      data: null
    });
  });
});
