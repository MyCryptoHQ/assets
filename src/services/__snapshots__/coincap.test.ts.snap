// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchCoinCapId initializes the cache 1`] = `
Array [
  Object {
    "id": "ethereum",
    "name": "Ethereum",
    "symbol": "ETH",
  },
]
`;
